feat: document Deep Agents Code plugins (#4905)
langchain-ai/docs: src/oss/deepagents/code/plugins.mdx

---

Plugins extend Deep Agents Code with reusable [skills](/oss/deepagents/code/memory-and-skills) and [MCP servers](/oss/deepagents/code/mcp-tools). Marketplaces provide catalogs for discovering and installing plugins across projects or teams. Deep Agents Code supports Claude- and Codex-style plugin manifests and marketplace catalogs, as described in [Create a plugin](#create-a-plugin) and [Create a marketplace](#create-a-marketplace).

<Warning>
    Install plugins and marketplaces only from sources you trust. An enabled plugin can add instructions and start MCP server processes with your user permissions.
</Warning>

## Manage plugins interactively

To browse marketplaces and manage plugins in a `dcode` session:

1. Run `/plugins` to open the plugin manager.
2. Add a marketplace from its **Marketplaces** tab. Supported sources include:
    - A GitHub repository in `owner/repo` format, optionally followed by `@branch-or-tag`.
    - An HTTPS Git repository URL, optionally followed by `#branch-or-tag`.
    - An HTTPS URL that serves a marketplace JSON file.
    - A local marketplace directory or JSON file.
3. Install a plugin from the marketplace.

> 4. Run `/reload` to activate newly installed plugin skills and MCP servers without restarting the session.

Reviewer: this seems unnecessary to mention since we show an action modal to users

The plugin manager also lets you enable, disable, and uninstall installed plugins. Disabling a plugin keeps it installed but excludes its skills and MCP servers after you run `/reload` or start a new session.

> Removing a marketplace uninstalls its plugins and removes managed cache data. Deep Agents Code preserves the original source when the marketplace came from a local directory or file. Run `/reload` or start a new session to apply the removal to an active session.

Reviewer: ```suggestion
Removing a marketplace uninstalls its plugins. Deep Agents Code preserves the original source when the marketplace came from a local directory or file.
```

## Manage plugins from the command line

Use `dcode plugin` for scripts and terminal-based administration. Plugin IDs use the format `plugin-name@marketplace-name`.

```bash
# Add and inspect a marketplace
dcode plugin marketplace add acme/plugins
dcode plugin marketplace list

# Browse and install plugins
dcode plugin list
dcode plugin install code-review@acme-tools

# Change plugin state
dcode plugin disable code-review@acme-tools
dcode plugin enable code-review@acme-tools

# Remove a plugin or marketplace
dcode plugin uninstall code-review@acme-tools
dcode plugin marketplace remove acme-tools
```

> `plugin list` and `plugin marketplace list` accept `--json`. After installing a plugin, run `/reload` in an active interactive session or start a new session.

Reviewer: > `plugin list` and `plugin marketplace list` accept `--json`

most people don't care and if they do this is noted in the `--help` option right?

if we're gonna keep it, mention it "to support machine readable output" or similar

## Use plugin skills and MCP servers

Plugin skills are namespaced to prevent collisions with project, user, and other plugin skills. Invoke a skill with its plugin ID and skill path:

```text
/skill:plugin-name@marketplace-name:skill-name optional arguments
```

In interactive mode, autocomplete also matches the shorter `/plugin-name:skill-name` form and expands it to the canonical `/skill:` command. Nested skill directories add each directory to the namespace. For example, `skills/review/security/SKILL.md` from `quality@acme-tools` becomes `/skill:quality@acme-tools:review:security`.

An enabled plugin can also contribute MCP servers. Deep Agents Code merges these servers with your regular MCP configuration when plugins load. Use `/mcp` to inspect available servers and tools.

## Create a plugin

> A Deep Agents Code plugin is a directory containing one or both of the supported components:

Reviewer: ```suggestion
A plugin is a directory containing one or both of the supported components:
```

```text
my-plugin/
├── .claude-plugin/
│   └── plugin.json
├── skills/
│   └── review/
│       └── SKILL.md
└── .mcp.json
```

> Deep Agents Code also recognizes `.codex-plugin/plugin.json`. The manifest is optional when components use their default locations. If the plugin contains one skill only, you can place `SKILL.md` at the plugin root instead of creating `skills/`.

Reviewer: is this an open spec? let's link to it

---
title: Plugins and marketplaces
sidebarTitle: Plugins
description: Install plugins from marketplaces or package skills and MCP servers for Deep Agents Code
---

> Plugins extend Deep Agents Code with reusable [skills](/oss/deepagents/code/memory-and-skills) and [MCP servers](/oss/deepagents/code/mcp-tools). Marketplaces provide catalogs for discovering and installing plugins across projects or teams. Deep Agents Code supports the Claude-compatible marketplace format described in [Create a marketplace](#create-a-marketplace).

Reviewer 2: Any reason we're not mentioning Codex-style marketplaces?
We ought to!

<Note>

>     Plugin support is experimental. Set `DEEPAGENTS_CODE_EXPERIMENTAL=1` before starting Deep Agents Code. The plugin format and commands may change.

Author: yes we are aware of this, this docs pr is a prerequisite to merge pr #4554

>     Plugin support is experimental. Set `DEEPAGENTS_CODE_EXPERIMENTAL=1` before starting Deep Agents Code. The plugin format and commands may change.

Reviewer 2: resolving this

</Note>

> 

Reviewer 2: ```suggestion

```

Reviewer 2: Eagerly suggesting removal

<Warning>
    Install plugins and marketplaces only from sources you trust. An enabled plugin can add instructions and start MCP server processes with your user permissions.
</Warning>

## Manage plugins interactively

> To browse marketplaces and manage plugins in a Deep Agents Code session:

Reviewer 2: ```suggestion
To browse marketplaces and manage plugins in a `dcode` session:
```

Reviewer 2: Let's start shifting nomenclature when folks know they're dealing with Deep Agents Code

wdyt @mdrxy

1. Set the experimental feature flag and start Deep Agents Code:

    ```bash
    export DEEPAGENTS_CODE_EXPERIMENTAL=1
    dcode
    ```

2. Run `/plugins` to open the plugin manager.
3. Add a marketplace from its **Marketplaces** tab. Supported sources include:

>     - A GitHub repository in `owner/repo` format, optionally followed by `@branch-or-tag`.

Reviewer: is the a marketplace spec? how does this work

    - An HTTPS Git repository URL, optionally followed by `#branch-or-tag`.
    - An HTTPS URL that serves a marketplace JSON file.
    - A local marketplace directory or JSON file.
4. Install a plugin from the marketplace.
5. Run `/reload` to activate newly installed plugin skills and MCP servers without restarting the session.

> The plugin manager also lets you enable, disable, and uninstall installed plugins. Disabling a plugin keeps it installed but excludes its skills and MCP servers the next time plugins load.

Reviewer 2: Maybe worth including a note that this requires a reload

> Removing a marketplace uninstalls its plugins and removes managed cache data. Deep Agents Code preserves the original source when the marketplace came from a local directory or file.

Reviewer 2: Does this require a reload to take effect?

Reviewer 2: Maybe if we built a "Reload plugins for changes to take effect" dialog into the exit pathway we woudln't need to document these things

But that's a Not Right Now task

## Manage plugins from the command line

Use `dcode plugin` for scripts and terminal-based administration. Plugin IDs use the format `plugin-name@marketplace-name`.

```bash
# Add and inspect a marketplace
dcode plugin marketplace add acme/plugins
dcode plugin marketplace list

# Browse and install plugins
dcode plugin list
dcode plugin install code-review@acme-tools

# Change plugin state
dcode plugin disable code-review@acme-tools
dcode plugin enable code-review@acme-tools

# Remove a plugin or marketplace
dcode plugin uninstall code-review@acme-tools
dcode plugin marketplace remove acme-tools
```

`plugin list` and `plugin marketplace list` accept `--json`. After installing a plugin, run `/reload` in an active interactive session or start a new session.

## Use plugin skills and MCP servers

Plugin skills are namespaced to prevent collisions with project, user, and other plugin skills. Invoke a skill with the plugin name followed by its path:

```text

> /plugin-name:skill-name optional arguments

Reviewer 2: Maybe I'm just having a logical disconnect but in `dcode` I use `/skill:{skill-name}`

Reviewer 2: I see this is mentioned in an openswe comment - while it does fuzzy match this still feels incorrect

```

> Nested skill directories add each directory to the namespace. For example, `skills/review/security/SKILL.md` in a plugin named `quality` becomes `/quality:review:security`.

Author: this isn't correct, can you please check the reference pr again, we fuzzy match

An enabled plugin can also contribute MCP servers. Deep Agents Code merges these servers with your regular MCP configuration when plugins load. Use `/mcp` to inspect available servers and tools.

## Create a plugin

A Deep Agents Code plugin is a directory containing one or both of the supported components:

```text
my-plugin/

> ├── .claude-plugin/

Reviewer 2: I wonder how we handle this longterm. It feels very weird using `.claude-plugin`. just something to keep thinking about
